atc/exec/task_step.go

```diff
@@ -157,8 +157,15 @@ func (step *TaskStep) run(ctx context.Context, state RunState) error {
 		// external task - construct a source which reads it from file
 		taskConfigSource = FileConfigSource{ConfigPath: step.plan.ConfigPath, Client: step.workerClient}
 
-		// for interpolation - use 'vars' from the pipeline, and then fill remaining with cred variables
-		taskVars = []vars.Variables{vars.StaticVariables(step.plan.Vars), variables}
+		// for interpolation - use 'vars' from the pipeline, and then fill remaining with cred variables.
+		// this 2-phase strategy allows to interpolate 'vars' by cred variables.
+		if len(step.plan.Vars) > 0 {
+			taskConfigSource = InterpolateTemplateConfigSource{
+				ConfigSource: taskConfigSource,
+				Vars:         []vars.Variables{vars.StaticVariables(step.plan.Vars)},
```

Comment:
I'm just trying to understand what's going on exactly so bear with me (this part of the code is pretty new to me):

Suppose you had the following pipeline:

```yaml
...
jobs:
- name: job
  plan:
  ...
  - load_var: foo
    file: some-resource/foo
  - task: run
    file: some-other-resource/task.yml
    vars:
      foo: ((.:foo))
```

where `some-other-resource/task.yml` was something like:

```yaml
...
run:
  path: echo
  args:
  - ((foo))
```

---

Am I correct in thinking that this first phase of interpolation result in the following `task.yml`:

```yaml
...
run:
  path: echo
  args:
  - ((.:foo))
```

and now we're able to interpolate `((.:foo))` using `step.delegate.Variables()`? Or am I misunderstanding the flow here?

Reply:
Yes, your understanding is right. Actually, I thought over the solution. The other solution I tried initially was to directly interpolate at the first place, which also worked, but introduced more code changes, thus needed more changes on unit tests. Then I figured out this way is simpler.

Reply:
Makes sense, clever solution!

Could you add a release note when you get the chance? Then I'm happy to approve the changes

Reply:
@aoldershaw I have just added a release note.